File: python/elizaos_plugin_mysticism/engines/iching.py
"""I Ching divination engine — Python port of typescript/src/engines/iching/.

Uses the three-coin method with cryptographic randomness (``secrets``).
All reading state is external (``IChingReadingState``); the engine is stateless.
"""

from __future__ import annotations

import json
import secrets
from pathlib import Path
from typing import Optional

from elizaos_plugin_mysticism.types import (
    CastResult,
    FeedbackEntry,
    Hexagram,
    HexagramLine,
    IChingReadingState,
    Trigram,
)

# ---------------------------------------------------------------------------
# Data loading
# ---------------------------------------------------------------------------

_DATA_DIR = Path(__file__).resolve().parent.parent.parent.parent / "data"

_trigrams_cache: Optional[list[Trigram]] = None
_hexagrams_cache: Optional[list[Hexagram]] = None


def _load_trigrams() -> list[Trigram]:
    global _trigrams_cache
    if _trigrams_cache is not None:
        return _trigrams_cache
    path = _DATA_DIR / "iching" / "trigrams.json"
    with open(path, "r", encoding="utf-8") as f:
        raw = json.load(f)
    _trigrams_cache = [Trigram(**t) for t in raw]
    return _trigrams_cache


def _load_hexagrams() -> list[Hexagram]:
    global _hexagrams_cache
    if _hexagrams_cache is not None:
        return _hexagrams_cache
    path = _DATA_DIR / "iching" / "hexagrams.json"
    with open(path, "r", encoding="utf-8") as f:
        raw = json.load(f)
    result: list[Hexagram] = []
    for h in raw:
        lines = [HexagramLine(**ln) for ln in h["lines"]]
        result.append(
            Hexagram(
                number=h["number"],
                name=h["name"],
                englishName=h["englishName"],
                character=h["character"],
                binary=h["binary"],
                topTrigram=h["topTrigram"],
                bottomTrigram=h["bottomTrigram"],
                judgment=h["judgment"],
                image=h["image"],
                lines=lines,
                keywords=h["keywords"],
                description=h["description"],
            )
        )
    _hexagrams_cache = result
    return _hexagrams_cache


# ---------------------------------------------------------------------------
# Lookup maps (lazily built)
# ---------------------------------------------------------------------------

_trigram_by_number: Optional[dict[int, Trigram]] = None
_hexagram_by_number: Optional[dict[int, Hexagram]] = None
_binary_to_number: Optional[dict[str, int]] = None


def _build_lookups() -> None:
    global _trigram_by_number, _hexagram_by_number, _binary_to_number
    if _trigram_by_number is not None:
        return
    trigs = _load_trigrams()
    hexes = _load_hexagrams()
    _trigram_by_number = {t.number: t for t in trigs}
    _hexagram_by_number = {h.number: h for h in hexes}
    _binary_to_number = {h.binary: h.number for h in hexes}


# ---------------------------------------------------------------------------
# Coin toss helpers
# ---------------------------------------------------------------------------


def _flip_coin() -> bool:
    """Return True (heads) or False (tails) with cryptographic randomness."""
    return secrets.randbelow(2) >= 1


def _cast_line() -> tuple[int, bool]:
    """Cast a single I Ching line using the three-coin method.

    Heads = 3, Tails = 2.
    Sum determines line type:
      6 (2+2+2) = Old Yin   (changing broken line)
      7 (2+2+3) = Young Yang (stable solid line)
      8 (2+3+3) = Young Yin  (stable broken line)
      9 (3+3+3) = Old Yang   (changing solid line)

    Returns:
        (value, changing) tuple.
    """
    coin1 = 3 if _flip_coin() else 2
    coin2 = 3 if _flip_coin() else 2
    coin3 = 3 if _flip_coin() else 2
    value = coin1 + coin2 + coin3
    return value, (value == 6 or value == 9)


def _line_value_to_binary(value: int) -> int:
    """7 and 9 are yang (solid) -> 1; 6 and 8 are yin (broken) -> 0."""
    return 1 if value in (7, 9) else 0


def _line_value_to_transformed_binary(value: int) -> int:
    """Transform changing lines: Old Yin (6) -> Yang (1), Old Yang (9) -> Yin (0)."""
    if value == 6:
        return 1  # Old Yin -> Yang
    if value == 9:
        return 0  # Old Yang -> Yin
    return _line_value_to_binary(value)


# ---------------------------------------------------------------------------
# Public casting functions
# ---------------------------------------------------------------------------


def cast_hexagram() -> CastResult:
    """Cast a full hexagram (6 lines, bottom to top)."""
    _build_lookups()

    cast_lines: list[tuple[int, bool]] = []
    for _ in range(6):
        cast_lines.append(_cast_line())

    lines = [cl[0] for cl in cast_lines]
    changing_lines = [i + 1 for i, cl in enumerate(cast_lines) if cl[1]]

    # Build binary string (bottom to top = left to right)
    binary = "".join(str(_line_value_to_binary(cl[0])) for cl in cast_lines)
    hexagram_number = binary_to_hexagram_number(binary)

    transformed_hexagram_number: Optional[int] = None
    transformed_binary: Optional[str] = None

    if changing_lines:
        transformed_binary = "".join(
            str(_line_value_to_transformed_binary(cl[0])) for cl in cast_lines
        )
        transformed_hexagram_number = binary_to_hexagram_number(transformed_binary)

    return CastResult(
        lines=lines,
        changingLines=changing_lines,
        hexagramNumber=hexagram_number,
        transformedHexagramNumber=transformed_hexagram_number,
        binary=binary,
        transformedBinary=transformed_binary,
    )


def binary_to_hexagram_number(binary: str) -> int:
    """Look up hexagram number from its 6-digit binary representation."""
    _build_lookups()
    assert _binary_to_number is not None
    number = _binary_to_number.get(binary)
    if number is None:
        raise ValueError(f"Unknown hexagram binary pattern: {binary}")
    return number


def get_hexagram(number: int) -> Hexagram:
    """Look up a hexagram by its King Wen number (1-64)."""
    _build_lookups()
    assert _hexagram_by_number is not None
    hexagram = _hexagram_by_number.get(number)
    if hexagram is None:
        raise ValueError(f"Hexagram number {number} not found (valid range: 1-64)")
    return hexagram


def get_trigram(number: int) -> Trigram:
    """Look up a trigram by number (1-8)."""
    _build_lookups()
    assert _trigram_by_number is not None
    trigram = _trigram_by_number.get(number)
    if trigram is None:
        raise ValueError(f"Trigram number {number} not found (valid range: 1-8)")
    return trigram


def get_lower_trigram(hexagram: Hexagram) -> Trigram:
    return get_trigram(hexagram.bottomTrigram)


def get_upper_trigram(hexagram: Hexagram) -> Trigram:
    return get_trigram(hexagram.topTrigram)


# ---------------------------------------------------------------------------
# IChingEngine
# ---------------------------------------------------------------------------


class IChingEngine:
    """Stateless I Ching engine — all reading state lives in IChingReadingState."""

    def start_reading(self, question: str) -> IChingReadingState:
        cast_result = cast_hexagram()
        hexagram = get_hexagram(cast_result.hexagramNumber)

        transformed_hexagram: Optional[Hexagram] = None
        if cast_result.transformedHexagramNumber is not None:
            transformed_hexagram = get_hexagram(cast_result.transformedHexagramNumber)

        return IChingReadingState(
            question=question,
            castResult=cast_result,
            hexagram=hexagram,
            transformedHexagram=transformed_hexagram,
            revealedLines=0,
            userFeedback=[],
        )

    def get_next_reveal(
        self,
        state: IChingReadingState,
    ) -> Optional[dict]:
        """Return the next changing line to reveal, or None when done."""
        sorted_changing = sorted(state.castResult.changingLines)

        if state.revealedLines >= len(sorted_changing):
            return None

        line_position = sorted_changing[state.revealedLines]
        return {"linePosition": line_position}

    def record_feedback(
        self,
        state: IChingReadingState,
        feedback: FeedbackEntry,
    ) -> IChingReadingState:
        return IChingReadingState(
            question=state.question,
            castResult=state.castResult,
            hexagram=state.hexagram,
            transformedHexagram=state.transformedHexagram,
            revealedLines=state.revealedLines + 1,
            userFeedback=[*state.userFeedback, feedback],
        )

    def get_synthesis(self, state: IChingReadingState) -> dict:
        """Return a summary of the reading."""
        return {
            "hexagram": {
                "number": state.hexagram.number,
                "name": state.hexagram.name,
                "englishName": state.hexagram.englishName,
            },
            "transformedHexagram": (
                {
                    "number": state.transformedHexagram.number,
                    "name": state.transformedHexagram.name,
                    "englishName": state.transformedHexagram.englishName,
                }
                if state.transformedHexagram
                else None
            ),
            "changingLines": state.castResult.changingLines,
            "question": state.question,
            "feedback": [
                {"element": fb.element, "userText": fb.userText}
                for fb in state.userFeedback
            ],
        }

    def is_complete(self, state: IChingReadingState) -> bool:
        sorted_changing = sorted(state.castResult.changingLines)
        return state.revealedLines >= len(sorted_changing)

    def get_casting_summary(self, state: IChingReadingState) -> str:
        """Human-readable summary of the cast."""
        hexagram = state.hexagram
        cast_result = state.castResult
        transformed = state.transformedHexagram

        upper = get_upper_trigram(hexagram)
        lower = get_lower_trigram(hexagram)

        parts: list[str] = [
            f"{hexagram.character} Hexagram {hexagram.number}: "
            f"{hexagram.name} — {hexagram.englishName}",
            "",
            f"Upper: {upper.character} {upper.englishName} ({upper.image})",
            f"Lower: {lower.character} {lower.englishName} ({lower.image})",
        ]

        if cast_result.changingLines:
            lines_str = ", ".join(f"Line {l}" for l in cast_result.changingLines)
            parts.extend(["", f"Changing lines: {lines_str}"])
        else:
            parts.extend(["", "No changing lines — the reading is stable."])

        if transformed:
            parts.extend([
                "",
                f"Transforming to: {transformed.character} Hexagram "
                f"{transformed.number}: {transformed.name} — {transformed.englishName}",
            ])

        return "\n".join(parts)

File: rust/src/engines/iching.rs
use std::collections::HashMap;

use rand::Rng;

use crate::types::{CastResult, Hexagram, Trigram};

// ---------------------------------------------------------------------------
// Static data loaded at compile time
// ---------------------------------------------------------------------------

const HEXAGRAMS_JSON: &str = include_str!("../../../data/iching/hexagrams.json");
const TRIGRAMS_JSON: &str = include_str!("../../../data/iching/trigrams.json");

fn load_hexagrams() -> Vec<Hexagram> {
    serde_json::from_str(HEXAGRAMS_JSON).expect("Failed to parse hexagrams.json")
}

fn load_trigrams() -> Vec<Trigram> {
    serde_json::from_str(TRIGRAMS_JSON).expect("Failed to parse trigrams.json")
}

// ---------------------------------------------------------------------------
// Line helpers
// ---------------------------------------------------------------------------

/// Three-coin toss result for a single line.
struct CastLineResult {
    /// Raw coin sum: 6, 7, 8, or 9
    value: u8,
    /// Whether this line is a changing line
    changing: bool,
}

/// Three coins are tossed.  Heads = 3, Tails = 2.
/// Sum determines line type:
///   6 (2+2+2) = Old Yin   — changing broken line
///   7 (2+2+3) = Young Yang — stable solid line
///   8 (2+3+3) = Young Yin  — stable broken line
///   9 (3+3+3) = Old Yang   — changing solid line
fn cast_line() -> CastLineResult {
    let mut rng = rand::thread_rng();
    let coin = |rng: &mut rand::rngs::ThreadRng| -> u8 {
        if rng.gen_bool(0.5) { 3 } else { 2 }
    };
    let c1 = coin(&mut rng);
    let c2 = coin(&mut rng);
    let c3 = coin(&mut rng);
    let value = c1 + c2 + c3;

    CastLineResult {
        value,
        changing: value == 6 || value == 9,
    }
}

/// Map a line value to its binary digit.
/// 7 and 9 are yang (solid) → 1
/// 6 and 8 are yin (broken) → 0
fn line_value_to_binary(value: u8) -> u8 {
    if value == 7 || value == 9 { 1 } else { 0 }
}

/// Map a line value to its *transformed* binary digit.
/// Old Yin (6) → Yang (1)
/// Old Yang (9) → Yin (0)
/// Young lines stay the same.
fn line_value_to_transformed_binary(value: u8) -> u8 {
    match value {
        6 => 1, // Old Yin → Yang
        9 => 0, // Old Yang → Yin
        _ => line_value_to_binary(value),
    }
}

// ---------------------------------------------------------------------------
// Public free functions
// ---------------------------------------------------------------------------

/// Cast a full hexagram using the three-coin method.
/// Lines are cast from bottom (position 1) to top (position 6).
pub fn cast_hexagram() -> CastResult {
    let cast_lines: Vec<CastLineResult> = (0..6).map(|_| cast_line()).collect();

    let lines: Vec<u8> = cast_lines.iter().map(|cl| cl.value).collect();
    let changing_lines: Vec<usize> = cast_lines
        .iter()
        .enumerate()
        .filter_map(|(i, cl)| if cl.changing { Some(i + 1) } else { None })
        .collect();

    // Build binary string (bottom to top = left to right)
    let binary: String = cast_lines
        .iter()
        .map(|cl| line_value_to_binary(cl.value).to_string())
        .collect();

    let hexagrams = load_hexagrams();
    let binary_to_number: HashMap<String, u32> =
        hexagrams.iter().map(|h| (h.binary.clone(), h.number)).collect();

    let hexagram_number = *binary_to_number
        .get(&binary)
        .unwrap_or_else(|| panic!("Unknown hexagram binary pattern: {}", binary));

    let (transformed_hexagram_number, transformed_binary) = if !changing_lines.is_empty() {
        let tb: String = cast_lines
            .iter()
            .map(|cl| line_value_to_transformed_binary(cl.value).to_string())
            .collect();
        let tn = *binary_to_number
            .get(&tb)
            .unwrap_or_else(|| panic!("Unknown hexagram binary pattern: {}", tb));
        (Some(tn), Some(tb))
    } else {
        (None, None)
    };

    CastResult {
        lines,
        changing_lines,
        hexagram_number,
        transformed_hexagram_number,
        binary,
        transformed_binary,
    }
}

/// Convert a binary string (e.g. "111111") to a hexagram number.
pub fn binary_to_hexagram_number(binary: &str) -> Result<u32, String> {
    let hexagrams = load_hexagrams();
    hexagrams
        .iter()
        .find(|h| h.binary == binary)
        .map(|h| h.number)
        .ok_or_else(|| format!("Unknown hexagram binary pattern: {}", binary))
}

/// Get a hexagram by its King Wen sequence number (1–64).
pub fn get_hexagram(number: u32) -> Result<Hexagram, String> {
    let hexagrams = load_hexagrams();
    hexagrams
        .into_iter()
        .find(|h| h.number == number)
        .ok_or_else(|| format!("Hexagram number {} not found (valid range: 1-64)", number))
}

/// Get a trigram by its number (1–8).
pub fn get_trigram(number: u32) -> Result<Trigram, String> {
    let trigrams = load_trigrams();
    trigrams
        .into_iter()
        .find(|t| t.number == number)
        .ok_or_else(|| format!("Trigram number {} not found (valid range: 1-8)", number))
}

/// Get the lower (bottom) trigram of a hexagram.
pub fn get_lower_trigram(hexagram: &Hexagram) -> Result<Trigram, String> {
    get_trigram(hexagram.bottom_trigram)
}

/// Get the upper (top) trigram of a hexagram.
pub fn get_upper_trigram(hexagram: &Hexagram) -> Result<Trigram, String> {
    get_trigram(hexagram.top_trigram)
}

// ---------------------------------------------------------------------------
// IChingEngine — stateful wrapper
// ---------------------------------------------------------------------------

pub struct IChingEngine {
    hexagrams: Vec<Hexagram>,
    trigrams: Vec<Trigram>,
    binary_to_number: HashMap<String, u32>,
}

impl IChingEngine {
    pub fn new() -> Self {
        let hexagrams = load_hexagrams();
        let trigrams = load_trigrams();
        let binary_to_number: HashMap<String, u32> =
            hexagrams.iter().map(|h| (h.binary.clone(), h.number)).collect();

        Self {
            hexagrams,
            trigrams,
            binary_to_number,
        }
    }

    /// Cast a full hexagram using the three-coin method.
    pub fn cast_hexagram(&self) -> CastResult {
        let cast_lines: Vec<CastLineResult> = (0..6).map(|_| cast_line()).collect();

        let lines: Vec<u8> = cast_lines.iter().map(|cl| cl.value).collect();
        let changing_lines: Vec<usize> = cast_lines
            .iter()
            .enumerate()
            .filter_map(|(i, cl)| if cl.changing { Some(i + 1) } else { None })
            .collect();

        let binary: String = cast_lines
            .iter()
            .map(|cl| line_value_to_binary(cl.value).to_string())
            .collect();

        let hexagram_number = *self
            .binary_to_number
            .get(&binary)
            .unwrap_or_else(|| panic!("Unknown hexagram binary pattern: {}", binary));

        let (transformed_hexagram_number, transformed_binary) = if !changing_lines.is_empty() {
            let tb: String = cast_lines
                .iter()
                .map(|cl| line_value_to_transformed_binary(cl.value).to_string())
                .collect();
            let tn = *self
                .binary_to_number
                .get(&tb)
                .unwrap_or_else(|| panic!("Unknown hexagram binary pattern: {}", tb));
            (Some(tn), Some(tb))
        } else {
            (None, None)
        };

        CastResult {
            lines,
            changing_lines,
            hexagram_number,
            transformed_hexagram_number,
            binary,
            transformed_binary,
        }
    }

    /// Look up a hexagram by King Wen number.
    pub fn get_hexagram(&self, number: u32) -> Option<&Hexagram> {
        self.hexagrams.iter().find(|h| h.number == number)
    }

    /// Look up a trigram by number.
    pub fn get_trigram(&self, number: u32) -> Option<&Trigram> {
        self.trigrams.iter().find(|t| t.number == number)
    }

    /// Convert a binary pattern to a hexagram number.
    pub fn binary_to_hexagram_number(&self, binary: &str) -> Option<u32> {
        self.binary_to_number.get(binary).copied()
    }

    /// Get the lower (bottom) trigram of a hexagram.
    pub fn get_lower_trigram(&self, hexagram: &Hexagram) -> Option<&Trigram> {
        self.get_trigram(hexagram.bottom_trigram)
    }

    /// Get the upper (top) trigram of a hexagram.
    pub fn get_upper_trigram(&self, hexagram: &Hexagram) -> Option<&Trigram> {
        self.get_trigram(hexagram.top_trigram)
    }
}

impl Default for IChingEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_64_hexagrams() {
        let hexagrams = load_hexagrams();
        assert_eq!(hexagrams.len(), 64);
    }

    #[test]
    fn loads_8_trigrams() {
        let trigrams = load_trigrams();
        assert_eq!(trigrams.len(), 8);
    }

    #[test]
    fn cast_hexagram_produces_valid_result() {
        let result = cast_hexagram();
        assert_eq!(result.lines.len(), 6);
        for &line in &result.lines {
            assert!(
                line == 6 || line == 7 || line == 8 || line == 9,
                "Line value must be 6, 7, 8, or 9 — got {}",
                line
            );
        }
        assert!(
            (1..=64).contains(&result.hexagram_number),
            "Hexagram number must be 1-64"
        );
        assert_eq!(result.binary.len(), 6);
    }

    #[test]
    fn hexagram_1_is_qian() {
        let hex = get_hexagram(1).unwrap();
        assert_eq!(hex.name, "Qian");
        assert_eq!(hex.binary, "111111");
    }

    #[test]
    fn trigram_1_is_qian() {
        let tri = get_trigram(1).unwrap();
        assert_eq!(tri.name, "Qian");
        assert_eq!(tri.binary, "111");
    }

    #[test]
    fn binary_lookup() {
        let n = binary_to_hexagram_number("000000").unwrap();
        assert_eq!(n, 2); // Kun / The Receptive
    }

    #[test]
    fn engine_cast_hexagram() {
        let engine = IChingEngine::new();
        let result = engine.cast_hexagram();
        assert_eq!(result.lines.len(), 6);
        assert!((1..=64).contains(&result.hexagram_number));
    }
}
